### PR TITLE
[USM] Handle Istio binary path

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

* Adds configuration value for accepting Istio binary path.
* Adds path resolution for Istion monitoring support.

### Motivation

We currently try to determine whether a PID represents an Envoy process using this constant path: `/proc/PID/bin/envoy`.

We identified a situation where Envoy is being used by a symlink path that is different from this constant path.

For that reason I have introduced in this PR 2 things:
* we have decided to use a configuration value for the Envoy path, which will allow us to update the searched path and include PIDs that are Envoy processes.

* we have added path resolution in order to be able to find those cases by default.

### Additional Notes

**How to use the configuration value?**

* **Option 1:** Add the following environment variable: `DD_SERVICE_MONITORING_CONFIG_TLS_ISTIO_ENVOY_PATH="/path_to_envoy"`
* **Option 2:** Add to the configuration file as the following:
```service_monitoring_config:
  enabled: true
  tls:
    istio:
      enabled: true
      envoy_path: "path_to_envoy"
```

**Performance impact** 
* Max CPU:
No negative impact 
Screenshot 2024-07-25 at 17 57 25

* Average CPU
No negative impact 
Screenshot 2024-07-25 at 17 58 06

* RSS average 
No negative impact 
Screenshot 2024-07-25 at 18 01 15

* Max RSS average 
No negative impact
Screenshot 2024-07-25 at 18 02 21

**The image was tested on staging cluster:
No negative impact of the change was observed on the staging cluster.**

### Possible

 Drawbacks / Trade-offs

### CODE REVIEW

File: pkg/network/usm/istio.go [hunk 1]
```diff
@@ -285,11 +290,17 @@ func (m *istioMonitor) getEnvoyPath(pid uint32) string {
 	}
 
 	buffer = buffer[:n]
-	i := bytes.Index(buffer, envoyCmd)
+	envoyPath, err := os.Readlink(m.envoyCmd)
```
Comment (guyarb) [inline pkg/network/usm/istio.go:293]: that's wrong
you should build the path `/proc/<pid>/exe` and run readlink on that

Comment (amitslavin) [reply pkg/network/usm/istio.go]: changed

File: pkg/network/usm/istio.go [hunk 2]
```diff
@@ -267,41 +266,12 @@ func (m *istioMonitor) handleProcessExec(pid uint32) {
 //
 // /usr/local/bin/envoy -cetc/istio/proxy/envoy-rev.json ...
 func (m *istioMonitor) getEnvoyPath(pid uint32) string {
-	cmdlinePath := fmt.Sprintf("%s/%d/cmdline", m.procRoot, pid)
 	exePath := fmt.Sprintf("%s/%d/exe", m.procRoot, pid)
 
-	f, err := os.Open(cmdlinePath)
-	if err != nil {
-		// This can happen often in the context of ephemeral processes
+	envoyPath, err := utils.ResolveSymlink(exePath)
```
Comment (guyarb) [inline pkg/network/usm/istio.go:271]: don't forget to modify the documentation of the function

File: pkg/network/usm/istio.go [hunk 3]
```diff
@@ -19,14 +18,15 @@ import (
...
-
 	manager "github.com/DataDog/ebpf-manager"
 )
 
 const (
 	istioSslReadRetprobe  = "istio_uretprobe__SSL_read"
 	istioSslWriteRetprobe = "istio_uretprobe__SSL_write"
+	// envoyCmd represents the search term used for determining
+	// whether or not a given PID represents an Envoy process.
+	envoyCmd = "/bin/envoy"
```
Comment (guyarb) [inline pkg/network/usm/istio.go:29]: it should be the default value of the configuration option

File: pkg/network/usm/istio_test.go [hunk 4]
```diff
@@ -140,6 +154,11 @@ func TestIstioSync(t *testing.T) {
 // PID 2 -> Bash process
 // PID 3 -> Envoy process
 func createFakeProcFS(t *testing.T) (procRoot string) {
+	_, err := os.Stat("/usr/bin/busybox")
```
Comment (guyarb) [inline pkg/network/usm/istio_test.go:157]: I'd add documentation

**Review by guyarb — APPROVED** [review]

**Review by richl-dd — APPROVED** [review]

File: pkg/network/usm/utils/pathutils.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"time"
+)
+
+// ResolveSymlink returns the target path of the symbolic link specified by linkPath.
+// If the link target is relative, it returns the relative path without resolving it to an absolute path.
+// If the link target cannot be resolved immediately, it retries for a short period.
+func ResolveSymlink(linkPath string) (string, error) {
+	targetPath, err := os.Readlink(linkPath)
+	if err != nil {
+		// We receive the Exec event, /proc could be slow to update
```
Comment (hmahmood) [inline pkg/network/usm/utils/pathutils.go:21]: Comment does not seem relevant

**Review by hmahmood — APPROVED** [review]

**Review by brycekahle — APPROVED** [review]

**Pull request merged**